ArjunaJTA/jta/tests/classes/com/hp/mwtests/ts/jta/recovery/RecoveryMonitorTestDeadlock.java

```diff
@@ -0,0 +1,133 @@
+    private static final int TEST_TIMEOUT_S = 1;
+
+    @Test
+    public void testRecoveryMonitorWithSuccess() throws Exception {
+        final CountDownLatch countdown = new CountDownLatch(1);
+        // XAResource which suspends the recover call to be able to run suspend on recovery manager meanwhile
+        final XAResource sleepingXA = new XATestResource(XATestResource.OK_JNDI_NAME, false) {
+            @Override
+            public Xid[] recover(int i) throws XAException {
+                countdown.countDown();
+                try {
+                    log.fine(this + ": Recovering for " + TEST_TIMEOUT_S + " seconds");
+                    TimeUnit.SECONDS.sleep(TEST_TIMEOUT_S);
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                }
+                System.out.println(this + ": Recovered.");
+                return new Xid[0];
+            }
+        };
+        // resource helper to be registered to recovery manager for it to process the XAResource
+        final XAResourceRecoveryHelper xaResourceRecoveryHelper = new XAResourceRecoveryHelper() {
+            @Override
+            public boolean initialise(String p) throws Exception {
+                return true;
+            }
+            @Override
+            public XAResource[] getXAResources() throws Exception {
+                return new XAResource[] {sleepingXA};
+            }
+        };
+
+        RecoveryEnvironmentBean recoveryEnvironmentBean = recoveryPropertyManager.getRecoveryEnvironmentBean();
+        recoveryEnvironmentBean.setRecoveryBackoffPeriod(1); // use a short interval between passes
+        recoveryEnvironmentBean.setRecoveryListener(true); // configure the RecoveryMonitor
+
+        RecoveryManager manager = RecoveryManager.manager(RecoveryManager.DIRECT_MANAGEMENT);
+        XARecoveryModule recoveryModule = new XARecoveryModule();
+        recoveryModule.addXAResourceRecoveryHelper(xaResourceRecoveryHelper);
+        manager.addModule(recoveryModule); // we only need to test the XARecoveryModule
+        manager.startRecoveryManagerThread(); // start periodic recovery
+
+        String host = recoveryEnvironmentBean.getRecoveryAddress(); // the recovery listener host
+        String rcPort = String.valueOf(recoveryEnvironmentBean.getRecoveryPort()); // the recovery listener port
+
+        // running the recovery scan which will be delayed by sleepingXAResource
+        manager.scan(null);
+        countdown.await(TEST_TIMEOUT_S, TimeUnit.SECONDS);
+
+        // sending client SCAN request while there is still running the recovery process
+        // the timeout on socket is defined so small that the call immediately returns here
+        // but SCAN invocation is already happening inside of the Naryaana
+        RecoveryMonitor.main(new String[] {"-verbose", "-port", rcPort, "-host", host, "-timeout", "10"});
```

Comment:
You have triggered a scan via a socket with an SO timeout of 10 msecs which will leave a WorkerService running inside periodic recovery. Next you trigger a RecoveryManager suspend and shutdown in a background thread (and wait 5 seconds for that to complete) and finally you validate that the shutdown request isn't deadlocked.

This all makes sense but I am not clear about how the countdown hits zero without the await call timing out?

Reply:
the current countdown (based on your previous comment I will add there one more to not depend on timeout in recovery) hits zero as there is first call `manager.scan(null)` which starts recovery. Then the `@Test` waits on countdown to ensure the `XAResource.recover()` is in progress and that's where the countdown is set to zero. After that the `RecoveryMonitor.main` is called.